lib/Support/WinFunctions.cpp

```diff
@@ -77,6 +80,97 @@ HRESULT CoGetMalloc(DWORD dwMemContext, IMalloc **ppMalloc) {
   return S_OK;
 }
 
+HANDLE CreateFile2(_In_ LPCWSTR lpFileName, _In_ DWORD dwDesiredAccess,
+                   _In_ DWORD dwShareMode, _In_ DWORD dwCreationDisposition,
+                   _In_opt_ void *pCreateExParams) {
+  return CreateFileW(lpFileName, dwDesiredAccess, dwShareMode, pCreateExParams,
+                     dwCreationDisposition, FILE_ATTRIBUTE_NORMAL, nullptr);
+}
 
-#endif // _WIN32
+HANDLE CreateFileW(_In_ LPCWSTR lpFileName, _In_ DWORD dwDesiredAccess,
+                   _In_ DWORD dwShareMode, _In_opt_ void *lpSecurityAttributes,
+                   _In_ DWORD dwCreationDisposition,
+                   _In_ DWORD dwFlagsAndAttributes,
+                   _In_opt_ HANDLE hTemplateFile) {
+  CW2A pUtf8FileName(lpFileName);
+  size_t fd = -1;
+  int flags = 0;
+
+  if (dwDesiredAccess & GENERIC_WRITE)
+    if (dwDesiredAccess & GENERIC_READ)
+      flags |= O_RDWR;
+    else
+      flags |= O_WRONLY;
+  else // flags may be 0, open() demands something here. This is mostly harmless
+    flags |= O_RDONLY;
+
+  if (dwCreationDisposition == CREATE_ALWAYS)
+    flags |= (O_CREAT | O_TRUNC);
+  if (dwCreationDisposition == OPEN_ALWAYS)
+    flags |= O_CREAT;
+  else if (dwCreationDisposition == CREATE_NEW)
+    flags |= (O_CREAT | O_EXCL);
+  else if (dwCreationDisposition == TRUNCATE_EXISTING)
+    flags |= O_TRUNC;
+  // OPEN_EXISTING represents default open() behavior
+
+  if (lpSecurityAttributes || hTemplateFile)
+    return INVALID_HANDLE_VALUE; // implementation limitation
```

Comment:
+1 for detecting implementation limitations.

But wouldn't it be better to replace the if conditions with assertions? that should make it easier to find issues in the future (rather than chasing invalid handles).
e.g.
```cpp
assert(!lpSecurityAttributes && "security attributes not supported in CreateFileW yet");
assert(!hTemplateFile && "template file not supported in CreateFileW yet");
```

Reply:
also applies to the next 3 implementation limitations.

Reply:
Sure.